apps/web/src/app/[locale]/[username]/_components/dashboard/in-progress-tab.tsx

```diff
@@ -1,42 +1,80 @@
+import { prisma } from '@repo/db';
 import {
   Table,
   TableBody,
-  TableCaption,
   TableCell,
   TableHead,
   TableHeader,
   TableRow,
 } from '@repo/ui/components/table';
+import Link from 'next/link';
+import { getRelativeTime } from '~/utils/relativeTime';
+import { withUnstableCache } from '~/utils/withUnstableCache';
 
-const challengesList = [
-  {
-    challengeTitle: 'Extreme Challenge 1',
-    challengeStatus: 'In-Progress',
-  },
-  {
-    challengeTitle: 'Extreme Challenge 4',
-    challengeStatus: 'In-Progress',
-  },
-];
+export async function InProgressTab({ userId }: { userId: string }) {
+  const challenges = await withUnstableCache({
+    fn: getInProgressChallenges,
+    args: [userId],
+    keys: ['in-progress-challenges'],
+    tags: ['challenges'],
+  });
 
-export function InProgressTab() {
   return (
     <Table>
-      <TableCaption>A list of your recent challenges.</TableCaption>
       <TableHeader>
         <TableRow>
           <TableHead>Challenge</TableHead>
-          <TableHead>Status</TableHead>
+          <TableHead>Last Submission</TableHead>
         </TableRow>
       </TableHeader>
       <TableBody>
-        {challengesList.map((challengeTitle) => (
-          <TableRow key={challengeTitle.challengeTitle}>
-            <TableCell className="font-medium">{challengeTitle.challengeTitle}</TableCell>
-            <TableCell>{challengeTitle.challengeStatus}</TableCell>
+        {challenges.map((challenge) => (
+          <TableRow key={challenge.id}>
+            <TableCell className="font-medium underline">
+              <Link href={`/challenge/${challenge.id}`}>{challenge.name}</Link>
+            </TableCell>
+            <TableCell>{getRelativeTime(challenge.submission[0]!.createdAt)}</TableCell>
           </TableRow>
         ))}
       </TableBody>
     </Table>
   );
 }
+
+async function getInProgressChallenges(id: string) {
+  const challenges = await prisma.challenge.findMany({
+    where: {
+      AND: [
+        {
+          submission: {
+            none: {
+              userId: id,
+              isSuccessful: true,
+            },
+          },
+        },
+        // Make sure there is at least one submission
+        { submission: { some: { userId: id, isSuccessful: false } } },
```

Comment:
Is this needed?

The none part makes sure there is either no successful submissions, or no submissions at all. I'm not sure what value the AND adds here.

Reply:
Without the `AND` it returned all challenges that had no submissions by my user.

Reply:
Oh, you're totally right! Ugh, the Prisma syntax is such a pain to read 😂

Reply:
Yeah I was disappointed we couldn't just use a subquery for this. I guess we could just do `rawQuery` but this works too. 😛